### ISSUE TITLE
[MU4 Issue] The date macros ($d, $D, $M) for header/footer show the dates in US format (MM/DD/YY)

### ISSUE THREAD

**Issue by Jojo-Schmitz** [musescore/MuseScore#15437]
**Describe the bug**
The date macros ($d, $D, $M) for header/footer show the dates in US format (MM/DD/YY)

**Expected behavior**
Same as MuseScore 3 and before: obey MuseScore's language settings or the computer's regional settings

As an alternative (and the least thing to do) use the ISO format, YYYY-MM-DD

**Comment by krtschil** [comment]
Tried with 

- MU4 - Official: Language setting is German. Date format is also German
- Nightly: Language setting is English. Date format is German

OS: Linux

**Comment by Jojo-Schmitz** [comment]
Not on Windows 11.

**Comment by wolf404** [comment]
Confirmed on Windows 10 21H2 (19044.2251): even after explicitly setting MS4 to German the dates are still in US format.
Tested with official (MuseScore version (64-bit): 4.0.0-223472200, revision: github-musescore-musescore-5485621) and 27 Dec nightly (MuseScore version (64-bit): 4.0.0-223610304, revision: github-musescore-musescore-002fd32)

**Comment by krtschil** [comment]
Confirmed on Windows 10 too. Date is always in US format

**Comment by wolf404** [comment]
I see the code differences, but my cpp skills are too limited to draw the right conclusions from what I see...

_version 4:_
https://github.com/musescore/MuseScore/blob/9055e9730b89f56bb8a04303e94429c198accf3f/src/engraving/libmscore/page.cpp#L463

_version 3.6.2:_
https://github.com/musescore/MuseScore/blob/4566605d92467b0f5a36b3731b64150500e48583/libmscore/page.cpp#L357

**Comment by Jojo-Schmitz** [comment]
So an issue in MuseScore's own string lib?

**Comment by wolf404** [comment]
As said: I don't have a clue of cpp so I cannot judge....
But it might well be that the _Date_ object always and only uses the underlying C locale (en_US). As long as you are on an en_US system it does not matter for obvious reasons, but there must have been a reason to use _QDate_ and Qt localization previously.

**Comment by cbjeukendrup** [comment]
A bit of background: the reason that we are not using Qt anymore for this, is that we don't want our engraving library to depend on such a specific and huge framework as Qt, which is primarily used for building user interfaces. To mention one advantage: without this huge dependency, it becomes much easier to use our library in other applications, if anyone would ever want that. Another is that our engraving module is safe when a new major version of Qt is released, or if Qt suddenly quits its open source licenses, like they are already doing for LTS releases.
However, the "loss" of Qt inevitably has some disadvantages; in some places, its easy localisation system will be missed. The solution might be to write a wrapper for Qt, like we have done for QPainter; so in this case, we could create an interface like "ILocalizationProvider" which would be used for various localised formatting tasks, with an underlying Qt implementation, but this implementation is hidden from the rest of the engraving module.

**Comment by wolf404** [comment]
For me it would be sufficient to have a possibility to specify my preferred date & time format somewhere (e.g. close to the language setting in general preferences).
Actually that would be even better than localization because I could use universal ISO format...

**Comment by Jojo-Schmitz** [comment]
If not configurable or as per locale, ISO format is the only right and sensible choice. Also that, and only that, should be in used on musescore.com, i.e. by the backend.

**Comment by mattok** [comment]
After uprgading my scores to version 4, I'm being bitten by this bug too, with the wrong date being displayed (for my locale - day and month in wrong order). I would ideally like to see:

1. The application picks up the locale of the local OS, whatever that is. (I appreciate that this is probably the hard part, given that MuseScore works cross-platform - which is important - and so probably has to account for all possible local OS libraries for determing it. I'm imagining lots of `ifdefs`.)
2. By default, new scores pick up the locale of the application, and the locale gets stored into the new score file, affecting how it is formatted. (This way, if the file is sent to someone else, it displays exactly as it was authored. That's my preference for the default anyway, since it is least likely to catch someone out, but an alternative for the default could be for it to always pick up what the application locale is.)
3. The musescore file locale can be altered by the user. (A specific locale can be chosen, or it can be set to always pick up the application locale.)
4. The date format can be configured. My preference would be for placeholders representing [the "conversion specifiers" of `strftime`'s format string](https://en.cppreference.com/w/c/chrono/strftime#Format_string) to be added to the existing ones already available for use within the header and footer text input fields. `Y`, `y`, `m` and `d` at a minimum.

Of these, number 4 is the most important. With number 4, it is possible to write a date in a locale-independent way (e.g. Y-m-d), which would make it possible to work around the problem of determining the user's locale problem, which sounds like more of a challenge. It would hopefully be a simple partial fix.

**Comment by Jojo-Schmitz** [comment]
The choice of the US format is the worst of all...
For me this issue is a showstopper for updating my scores to 4.0, as I'm using `Version $M`on all my scores, and that US format is just confusing as hell to non-US users

**Comment by wolf404** [comment]
@cbjeukendrup : any chance to quickly change the default format to ISO (yyyy-MM-dd) instead of _LocaleShortFormat_? 
https://github.com/musescore/MuseScore/blob/9055e9730b89f56bb8a04303e94429c198accf3f/src/engraving/libmscore/page.cpp#L463

That would for sure be an excellent workaround for all non-US users - a really quick win with minimal effort.

**Comment by Jojo-Schmitz** [comment]
Yes, ISO would be the very least thing to do here. At least for a 4.0.2.
For 4.1 I'd rather see a more adcanced solution

**Comment by cbjeukendrup** [comment]
Let's try that. See #15906.

@Tantacrul This seems indeed appropriate for 4.0.2 (and is low risk anyway)

**Comment by cbjeukendrup** [comment]
Reopening awaiting a proper solution for 4.1

**Comment by Jojo-Schmitz** [comment]
Came up again in https://musescore.org/en/node/342785

**Comment by Jojo-Schmitz** [comment]
Hasn't this fix made it into 4.0.2?

**Comment by Tantacrul** [comment]
> Reopening awaiting a proper solution for 4.1

Based on @cbjeukendrup's comment, I'd say 'no'.

**Comment by Jojo-Schmitz** [comment]
That comment shouldn't indicate that, just that for 4.1 a more complete solution is wanted than the simple ISO date one for 4.0.2.
But instead the ISO version went into master but not into 4.0.2 :-(

**Comment by Tantacrul** [comment]
I must admit, this one was handled by other folk, so I think I'll wait for someone who was involved to comment. Apologies, just a little busy so can't quite take time to look.

**Comment by Tantacrul** [comment]
Ah, well sorry if this seems rather convoluted. It is entirely likely that someone on the dev team asked me to move it for some reason. At least if it is assigned to 4.1, we still have the task prepared to be worked on.

**Comment by Jojo-Schmitz** [comment]
Anything happening on this?

**Comment by grigorax** [comment]
I too am annoyed by the US format even though it has a logic of its own ...  Thanks for the background info, I now appreciate more the convolutions that go with this seemingly small issue.

**Comment by Jojo-Schmitz** [comment]
Came up again in https://musescore.org/en/node/368199

**Comment by Jojo-Schmitz** [comment]
This issue should get marked as a Mu3 regression

**Comment by GrahamDickins** [comment]
Thank you Jojo-Schmitz.

Just to echo that I have also been disappointed by this ongoing issue and I too would like to see an option to choose date format or use local (at least as default).

In my case, currently I use in a footer a static Creation Date string (dd/MM/yyyy, as per my regional standard, hailing from the UK) beside a dynamic Modified Date string ($M), but have to put up with the inconsistency that the latter invariably displays as yyyy-mm-dd.

I attach a clipped screenshot for example.
![{E17879EE-C46A-4522-A00C-6EB91962C5B1}](https://github.com/user-attachments/assets/900d7dcd-a5a4-4374-b2f5-4382bbfefc59)

I hope this can be resolved in the not-too distant future.

Kind regards,
Graham

PS version + revision paste: OS: Windows 11 Version 2009 or later, Arch.: x86_64, MuseScore Studio version (64-bit): 4.4.2-242570931, revision: github-musescore-musescore-3130f97

**Comment by Jojo-Schmitz** [comment]
>  (though in my case it's forcing ISO, rather than US) and I too would like to see an option to choose date format or use local (at least as default).

Yes, it got changed from US to ISO format, as a temporary quick and dirty workaround. But still the old locale-dependant format has not been reinstated.

**Comment by Jojo-Schmitz** [comment]
Came up once again in https://musescore.org/en/node/378660

**Comment by Jojo-Schmitz** [comment]
Came up again in https://musescore.org/en/node/386008

**Comment by Jojo-Schmitz** [comment]
As the muse string library now seems to have been move out into the muse_framework submodule, I guess an issue there is needed, in addition? Add there to provide the infrastructure, keep here to make use of it?

Hold on, seems there is `muse::DateFormat::LocaleShortFormat`, so infrastructure seems to exist, can't we just use that?

Nah, that apparently gives the (bogus and bad) US format for some strange reason, seems there's the real bug?

**Comment by Jojo-Schmitz** [comment]
The problem seems to be that a call to `std::setlocale(LC_TIME, "de")` (or whatever the current locale or language setting is or should be) is missing prior to calling `std::strftime()`.

Here's one way to fix it, in the muse_framework submodule, languagesconfiguration.cpp, line 53:
```cpp
ValCh<QString> LanguagesConfiguration::currentLanguageCode() const
{
    ValCh<QString> result;
    result.ch = m_currentLanguageCodeChanged;
    result.val = settings()->value(LANGUAGE_KEY).toQString();
    std::setlocale(LC_TIME, result.val.toStdString().c_str()); //<<<<<<<<<<<<< add this line

    return result;
}
```
and in headerfooterlayout.cpp replace all(?) `muse::DateFormat::ISODate` by `muse::DateFormat::LocaleShortFormat`.

(And on the backend make sure to set LC_TIME to a locale that uses the ISO dates/times, like e.g. "af" or "en_CA", as I found out by accident)

Would this be a viable thing to do?

See https://github.com/musescore/muse_framework/pull/79

**Comment by Jojo-Schmitz** [comment]
Seems on Discord we agreed on making it a (per score) style setting, possibly defaulting to the locale of the translation used at score creation.
That way every user can have it the way (s)he likes and it looks the same for everyone opening the score, including musescore.com
